entity-framework/core/querying/null-semantics.md

```diff
@@ -0,0 +1,131 @@
+
+```sql
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE ([e].[Id] <> [e].[NullableInt]) OR [e].[NullableInt] IS NULL
+```
+
+Queries four and five show the pattern when both columns are nullable. It's worth noting that `<>` operation produces significantly more complicated (and slower) query than equality operation.
+
+```sql
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE ([e].[String1] = [e].[String2]) OR ([e].[String1] IS NULL AND [e].[String2] IS NULL)
+
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE (([e].[String1] <> [e].[String2]) OR ([e].[String1] IS NULL OR [e].[String2] IS NULL)) AND ([e].[String1] IS NOT NULL OR [e].[String2] IS NOT NULL)
+```
+
+## Null semantics in functions
+
+Many functions in SQL can only return null result if some of their arguments are null. EF Core takes advantage of this to produce more efficient queries.
+The query below illustrates the optimization:
+
+[!code-csharp[Main](../../../samples/core/Querying/NullSemantics/Program.cs#Functions)]
+
+The generated SQL is as follows (we don't need to evaluate the `SUBSTRING` function):
+
+```sql
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE [e].[String1] IS NULL OR [e].[String2] IS NULL
+```
+
+This optimization can also be used for user defined functions. This can be done by adding `PropagatesNullability()` call to relevant function parameters model configuration.
+To illustrate this, define two user functions inside the `DbContext`:
+
+[!code-csharp[Main](../../../samples/core/Querying/NullSemantics/NullSemanticsContext.cs#UdfBody)]
+
+The model configuration (inside `OnModelCreating` method) is as follows:
+
+[!code-csharp[Main](../../../samples/core/Querying/NullSemantics/NullSemanticsContext.cs#UdfModelConfiguration)]
+
+First function is configured in a standard way and the second function is configured to take advantage of the nullability propagation optimization.
+
+When issuing the following queries:
+
+[!code-csharp[Main](../../../samples/core/Querying/NullSemantics/Program.cs#UdfExamples)]
+
+We get this SQL:
+
+```sql
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE [dbo].[ConcatStrings]([e].[String1], [e].[String2]) IS NOT NULL
+
+SELECT [e].[Id], [e].[Int], [e].[NullableInt], [e].[String1], [e].[String2]
+FROM [Entities] AS [e]
+WHERE [e].[String1] IS NOT NULL AND [e].[String2] IS NOT NULL
+```
+
+Just like with built-in `Substring` function, the second query doesn't need to evaluate the function itself to test it's nullability.
+
+> [!NOTE]
+> This optimization should only be used if the function can only return `null` becuase it's parameters are `null`.
+
+## Optimizations and considerations
```

Comment:
Any ideas for better name? Wanted to use "best practices" but acrolinx is complaining

Reply:
Write performant queries

Reply:
We should also link to this from [the perf docs on efficient querying](https://docs.microsoft.com/en-us/ef/core/performance/efficient-querying), possibly add an "Additional resources" section there to link to here.